cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,203 @@
+		&provider{
+			name:         "kubeadm",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/bootstrap-components.yaml",
+			providerType: clusterctlv1.BootstrapProviderType,
+		},
+	}
+
+	// ensure defaults are consistently sorted
+	sort.Slice(defaults, func(i, j int) bool {
+		return defaults[i].Name() < defaults[j].Name()
+	})
+
+	return defaults
+}
+
+func (p *providersClient) List() ([]Provider, error) {
+	// Creates a maps with all the defaults provider configurations
+	rMap := map[string]Provider{}
+	for _, r := range p.defaults() {
+		rMap[r.Name()] = r
+	}
+
+	// Gets user defined provider configurations, validate them, and merges with
+	// hard-coded configurations handling conflicts (user defined take precedence on hard-coded)
+	type configProvider struct {
+		Name string                    `json:"name,omitempty"`
+		URL  string                    `json:"url,omitempty"`
+		Type clusterctlv1.ProviderType `json:"type,omitempty"`
+	}
+
+	var ur []configProvider //nolint
+	err := p.reader.UnmarshalKey(ProvidersConfigKey, &ur)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal providers from the clusterctl configuration file")
+	}
+
+	for _, u := range ur {
+		r := NewProvider(u.Name, u.URL, u.Type)
+
+		err := validateProviderRepository(r)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error validating configuration from %q. Please fix the providers value in clusterctl configuration file", r.Name())
+		}
+
+		if _, ok := rMap[r.Name()]; ok {
+			klog.V(3).Infof("The clusterctl configuration file overrides default configuration for provider %s", r.Name())
+		}
+		rMap[r.Name()] = r
+	}
+
+	// Converts provider configurations maps to a slice
+	var providers []Provider //nolint
+	for _, r := range rMap {
+		providers = append(providers, r)
+	}
+
+	// ensure provider configurations are consistently sorted
+	sort.Slice(providers, func(i, j int) bool {
+		return providers[i].Name() < providers[j].Name()
+	})
+
+	return providers, nil
+}
+
+func (p *providersClient) Get(name string) (*Provider, error) {
+	l, err := p.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range l {
+		if name == r.Name() {
+			return &r, nil
+		}
+	}
+
+	return nil, errors.Errorf("failed to get configuration for %q provider. Please check the provider name and/or add configuration for new providers using the .clusterctl config file", name)
+}
+
+func validateProviderRepository(r Provider) error {
+	if r.Name() == "" {
+		return errors.New("name value cannot be empty")
+	}
+	errMsgs := validation.IsDNS1123Subdomain(r.Name())
+	if len(errMsgs) != 0 {
+		return errors.Errorf("invalid name value: %s", strings.Join(errMsgs, "; "))
+	}
+	if r.URL() == "" {
+		return errors.New("url value cannot be empty")
+	}
+
+	_, err := url.Parse(r.URL())
+	if err != nil {
```

Comment:
`if err := ...; err != nil`

Reply:
done

Reply:
The diff doesn't seem to have the change